src/plot.cpp

```diff
@@ -1,10 +1,1034 @@
+} // end draw_mesh_lines
+
+//===============================================================================
+// CREATE_VOXEL outputs a binary file that can be input into silomesh for 3D
+// geometry visualization.  It works the same way as create_ppm by dragging a
+// particle across the geometry for the specified number of voxels. The first
+// 3 int(4)'s in the binary are the number of x, y, and z voxels.  The next 3
+// real(8)'s are the widths of the voxels in the x, y, and z directions. The next
+// 3 real(8)'s are the x, y, and z coordinates of the lower left point. Finally
+// the binary is filled with entries of four int(4)'s each. Each 'row' in the
+// binary contains four int(4)'s: 3 for x,y,z position and 1 for cell or material
+// id.  For 1 million voxels this produces a file of approximately 15MB.
+//===============================================================================
+
+void create_voxel(ObjectPlot *pl)
+{
+
+  // compute voxel widths in each direction
+  double vox[3];
+  vox[0] = pl->width[0]/(double)pl->pixels[0];
+  vox[1] = pl->width[1]/(double)pl->pixels[1];
+  vox[2] = pl->width[2]/(double)pl->pixels[2];
+
+  // initial particle position
+  double ll[3];
+  ll[0] = pl->origin[0] - pl->width[0] / TWO;
+  ll[1] = pl->origin[1] - pl->width[1] / TWO;
+  ll[2] = pl->origin[2] - pl->width[2] / TWO;
+
+  // allocate and initialize particle
+  double dir[3] = {HALF, HALF, HALF};
+  Particle *p = new Particle();
+  p->initialize();
+  std::copy(ll, ll + 3, p->coord[0].xyz);
+  std::copy(dir, dir + 3, p->coord[0].uvw);
+  p->coord[0].universe = openmc_root_universe;
+
+  // Open binary plot file for writing
+  std::ofstream of;
+  std::string fname = std::string(pl->path_plot);
+  fname = strtrim(fname);
+  hid_t file_id = file_open(fname, 'w');
+
+  // write header info
+  write_attribute(file_id, "filetype", "voxel");
+  write_attribute(file_id, "version", VERSION_VOXEL);
+  write_attribute(file_id, "openmc_version", VERSION);
+
+#ifdef GIT_SHA1
+  write_attribute(file_id, "git_sha1", GIT_SHA1);
+#endif
+
+  // Write current date and time
+  write_attribute(file_id, "date_and_time", time_stamp().c_str());
+  hsize_t three = 3;
+  write_attr_int(file_id, 1, &three, "num_voxels", pl->pixels);
+  write_attr_double(file_id, 1, &three, "voxel_width", vox);
+  write_attr_double(file_id, 1, &three, "lower_left", ll);
+
+  // Create dataset for voxel data -- note that the dimensions are reversed
+  // since we want the order in the file to be z, y, x
+  hsize_t dims[3];
+  dims[0] = pl->pixels[0];
+  dims[1] = pl->pixels[1];
+  dims[2] = pl->pixels[2];
+  hid_t dspace, dset, memspace;
+  voxel_init(file_id, &(dims[0]), &dspace, &dset, &memspace);
+
+  // move to center of voxels
+  ll[0] = ll[0] + vox[0] / TWO;
+  ll[1] = ll[1] + vox[1] / TWO;
+  ll[2] = ll[2] + vox[2] / TWO;
+
+  int data[pl->pixels[1]][pl->pixels[2]];
+
+  int rgb[3], id;
+  for (int x = 0; x < pl->pixels[0]; x++) {
+    // progress bar here
```

Comment:
What was your plan for dealing with the progress bar?

Reply:
Don't have one yet. I'll look into that.